### PR TITLE
Transfer learning update

### PR DESCRIPTION
**Pull request by okntemujin**
Hello, I implemented the transfer learning model. I am not quite sure about the test_scalers, so I did them how I understood it.

### CODE REVIEW

File: multifidelity_end2end.py [hunk 1]
```diff
@@ -83,7 +83,22 @@ def main():
...
+        data_df, lf_train_smiles, hf_train_smiles = separate_hf_and_lf_train(data_df, train_smiles, args)
+        data_df, lf_val_smiles, hf_val_smiles = separate_hf_and_lf_train(data_df, val_smiles, args)
+        data_df, lf_test_smiles, hf_test_smiles = separate_hf_and_lf_train(data_df, test_smiles, args)
+
+        print("HF train size:", len(hf_train_smiles))
+        print("LF train size:", len(lf_train_smiles))
+        print("HF val size:", len(hf_val_smiles))
+        print("LF val size:", len(lf_val_smiles))
+        print("HF test size:", len(hf_test_smiles))
+        print("LF test size:", len(lf_test_smiles))
```
Comment (kevingreenman) [inline multifidelity_end2end.py]: can you also add the print statements of the relevant sizes for delta ML? I missed that in the last PR

File: multifidelity_end2end.py [hunk 2]
```diff
@@ -147,13 +207,21 @@ def main():
         devices=1,
         max_epochs=args.num_epochs,
     )
-    trainer.fit(mpnn, train_loader, val_loader)
 
-    preds = trainer.predict(mpnn, test_loader)
+    if args.model_type == "transfer":
+        
+        trainer.fit(mpnn, lf_train_loader, lf_val_loader)
```
Comment (kevingreenman) [inline multifidelity_end2end.py]: I think we should also do `trainer.predict` after the fitting on the LF data. We should do separate predictions on the `hf_test_loader` and `lf_test_loader`.

File: multifidelity_end2end.py [hunk 3]
```diff
@@ -147,13 +207,21 @@ def main():
...
+        trainer.fit(mpnn, hf_train_loader, hf_val_loader)
+        preds = trainer.predict(mpnn, hf_test_loader)
```
Comment (kevingreenman) [inline multifidelity_end2end.py]: We should also do predictions on the `lf_test_loader`.

File: multifidelity_end2end.py [hunk 4]
```diff
@@ -117,25 +141,61 @@ def main():
...
+        lf_test_dset = data.MoleculeDataset(lf_test_data, mgf)
 
-    train_loader = data.MolGraphDataLoader(train_dset, batch_size=50, num_workers=12)
-    val_loader = data.MolGraphDataLoader(val_dset, batch_size=50, num_workers=12, shuffle=False)
-    test_loader = data.MolGraphDataLoader(test_dset, batch_size=50, num_workers=12, shuffle=False)
+        if args.scale_data:
+            # Is this correct?
+            train_scaler = hf_train_dset.normalize_targets()
+            _ = hf_val_dset.normalize_targets(train_scaler)
+            test_scaler = hf_test_dset.normalize_targets()
```
Comment (kevingreenman) [inline multifidelity_end2end.py]: The LF data needs to be normalized too (separate scaler for LF)

Comment (kevingreenman) [inline multifidelity_end2end.py]: I'm not sure why I wrote it this way originally - the test set should also be scaled using the train scaler. Can you help me update the later parts of the script that use the `test_scaler` to use `train_scaler` instead?
```suggestion
            _ = hf_test_dset.normalize_targets(train_scaler)
```

File: multifidelity_end2end.py [hunk 5]
```diff
@@ -117,25 +141,61 @@ def main():
...
+
+    else:
+        train_dset = data.MoleculeDataset(train_data, mgf)
+        val_dset = data.MoleculeDataset(val_data, mgf)
+        test_dset = data.MoleculeDataset(test_data, mgf)
+
+        if args.scale_data:
+            train_scaler = train_dset.normalize_targets()
+            _ = val_dset.normalize_targets(train_scaler)
+            test_scaler = test_dset.normalize_targets()
```
Comment (kevingreenman) [inline multifidelity_end2end.py]: I'm not sure why I wrote it this way originally - the test set should also be scaled using the train scaler. Can you help me update the later parts of the script that use the `test_scaler` to use `train_scaler` instead?
```suggestion
            _ = test_dset.normalize_targets(train_scaler)
```

File: multifidelity_end2end.py (hunk 4 above)
Comment (kevingreenman) [reply multifidelity_end2end.py]: you'll probably have to modify later parts of the script to account for there being two separate scalers

**Review by kevingreenman — CHANGES_REQUESTED** [review]

File: multifidelity_end2end.py [hunk 6]
```diff
@@ -141,61 +145,50 @@ def main():
...
+
         hf_train_dset = data.MoleculeDataset(hf_train_data, mgf)
-        lf_train_dset = data.MoleculeDataset(lf_train_data, mgf)
         hf_val_dset = data.MoleculeDataset(hf_val_data, mgf)
-        lf_val_dset = data.MoleculeDataset(lf_val_data, mgf)
         hf_test_dset = data.MoleculeDataset(hf_test_data, mgf)
-        lf_test_dset = data.MoleculeDataset(lf_test_data, mgf)
 
         if args.scale_data:
             # Is this correct?
```
Comment (kevingreenman) [inline multifidelity_end2end.py]:
```suggestion
```

**Review by kevingreenman — APPROVED** [review]
